From d5925f5b614d2b8eca9f74d3ced59cbbb9521451 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Thu, 14 Jul 2022 16:50:08 +0200
Subject: [PATCH 0766/1010] evl/net: skbuff: clear linkage before sending
 in-band

We use the inner list head available from the skbuff header, which
lives in a union with a couple of open-coded list pointers. Make sure
to clear the latter before sending to the in-band stack, so that these
are not followed.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/net/output.c | 17 +++++++++++------
 1 file changed, 11 insertions(+), 6 deletions(-)

diff --git a/kernel/evl/net/output.c b/kernel/evl/net/output.c
index 70e928efc3..22ccd2a56e 100644
--- a/kernel/evl/net/output.c
+++ b/kernel/evl/net/output.c
@@ -101,6 +101,14 @@ void evl_net_do_tx(void *arg)
 	}
 }
 
+static void skb_xmit_inband(struct sk_buff *skb)
+{
+	evl_uncharge_socket_wmem(skb);
+	skb->prev = NULL;
+	skb->next = NULL;
+	dev_queue_xmit(skb);
+}
+
 /* in-band hook, called upon NET_TX_SOFTIRQ. */
 void skb_inband_xmit_backlog(void)
 {
@@ -108,10 +116,8 @@ void skb_inband_xmit_backlog(void)
 	LIST_HEAD(list);
 
 	if (evl_net_move_skb_queue(this_cpu_ptr(&oob_tx_relay), &list)) {
-		list_for_each_entry_safe(skb, n, &list, list) {
-			evl_uncharge_socket_wmem(skb);
-			dev_queue_xmit(skb);
-		}
+		list_for_each_entry_safe(skb, n, &list, list)
+			skb_xmit_inband(skb);
 	}
 }
 
@@ -188,8 +194,7 @@ int evl_net_transmit(struct sk_buff *skb) /* oob or in-band */
 	 * xmit_inband().
 	 */
 	if (running_inband()) {
-		evl_uncharge_socket_wmem(skb);
-		dev_queue_xmit(skb);
+		skb_xmit_inband(skb);
 		return 0;
 	}
 
-- 
2.30.2

